crates/sui-core/src/consensus_adapter.rs

```diff
@@ -312,28 +306,49 @@ impl ConsensusAdapter {
         } else {
             None
         };
-        self.authority
-            .database
-            .epoch_store()
-            .insert_pending_consensus_transactions(&transaction)?;
+        epoch_store.insert_pending_consensus_transactions(&transaction)?;
         if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
             self.pending_certificates
                 .lock()
                 .insert(*certificate.digest());
         }
-        Ok(self.submit_unchecked(transaction))
+        Ok(self.submit_unchecked(transaction, epoch_store.clone()))
     }
 
-    fn submit_unchecked(self: &Arc<Self>, transaction: ConsensusTransaction) -> JoinHandle<()> {
+    fn submit_unchecked(
+        self: &Arc<Self>,
+        transaction: ConsensusTransaction,
+        epoch_store: Arc<AuthorityPerEpochStore<AuthoritySignInfo>>,
+    ) -> JoinHandle<()> {
         // Reconfiguration lock is dropped when pending_consensus_transactions is persisted, before it is handled by consensus
-        let async_stage = self.clone().submit_and_wait(transaction);
+        let async_stage = self.clone().submit_and_wait(transaction, epoch_store);
         // Number of this tasks is limited by `sequencing_certificate_inflight` limit
         let join_handle = spawn_monitored_task!(async_stage);
         join_handle
     }
 
+    async fn submit_and_wait(
+        self: Arc<Self>,
+        transaction: ConsensusTransaction,
+        epoch_store: Arc<AuthorityPerEpochStore<AuthoritySignInfo>>,
+    ) {
+        let epoch_terminated = epoch_store.wait_epoch_terminated().boxed();
```

Comment:
I don't quite understand this one: does `wait_epoch_terminated` wait for reconfiguration to finish? If so, why do we need to wait for that each time we are submitting a transaction?

Reply:
It waits for this epoch to end. We don't wait for it on this line, only create a Future.
On the next line we `select` between completion of normal submission process or end of epoch(in which case submission is cancelled)